### PR TITLE
Feature/testing

### LINKED ISSUE #3: Add Spec Tests to Warduino
https://github.com/kanaka/wac#running-webassembly-spec-tests

### PR DESCRIPTION
**Pull request by tolauwae**
Closes #3

### CODE REVIEW

File: .github/workflows/compile.yml [hunk 1]
```diff
@@ -15,18 +15,86 @@ jobs:
...
+                      sketches: |
+                        - tests/compilation/esp32
+#                   - board:
+#                       platform-name: esp8266:esp8266
+#                     platforms: |
+                        # Install ESP8266 platform via Boards Manager
+#                       - name: esp8266:esp8266
+#                         source-url: https://arduino.esp8266.com/stable/package_esp8266com_index.json
+#                         version: 2.5.0
+#                    sketch-paths:
```
Comment (beardhatcode) [inline .github/workflows/compile.yml:57]: I would remove these commented lines

```suggestion
```

File: debug.h [hunk 2]
```diff
@@ -23,24 +23,44 @@
...
+#ifdef ARDUINO
+
 #define ASSERT(exp, ...)                                           \
     {                                                              \
         if (!(exp)) {                                              \
             printf("Assert Failed (%s:%d): ", __FILE__, __LINE__); \
             printf(__VA_ARGS__);                                   \
-            printf("\n\n\nlooping...\n");                          \
-            end();                                                 \
+            exit(-1);                                              \
```
Comment (beardhatcode) [inline debug.h:33]: if I recall correctly we had an inf loop here to ensure that the message could be read and that the serial output was not just spammed with restarts.

File: tests/integration/wasm_tests.cpp [hunk 3]
```diff
@@ -0,0 +1,355 @@
...
+        }
+    }
+}
+
+void invoke(Module *m, const char *call_f, Value values[]) {
+    uint32_t fidx = -1;
+    m->sp = -1;
+    m->fp = -1;
+    m->csp = -1;
+    // TODO move to the WARDuino class
```
Comment (beardhatcode) [inline tests/integration/wasm_tests.cpp:70]: This would indeed be a convenient addition to the WARDuino class

**Review by beardhatcode — APPROVED** [review]

File: tests/integration/wasm_tests.cpp (hunk 3 above)
Comment (tolauwae) [reply tests/integration/wasm_tests.cpp:70]: When we move this, it makes sense to immediately add support for multiple modules as well. Adding this to issue #27.

File: cli/main.cpp [hunk 4]
```diff
@@ -95,12 +107,25 @@ int main(int argc, const char *argv[]) {
...
     if (argc == 1) {
         ARGV_GET(file_name);
+        ARGV_SHIFT();
+    }
+
+    if (argc == 0 && file_name != nullptr) {
+        if (run_tests) {
+            run_wasm_test(wac, file_name, asserts_file, watcompiler);
+            return 0;
+        }
```
[the remark this answers is not in the document]
Comment (tolauwae) [reply cli/main.cpp:135]: commit 87729cb

**Pull request merged**